program/src/state.rs

```diff
@@ -3,58 +3,546 @@
 
 //! State transition types
 
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::marker::PhantomData;
 use std::ops::Range;
 
 use serde::Serialize;
 
 use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
-use solana_program::borsh::{get_instance_packed_len, try_from_slice_unchecked};
-use solana_program::clock::Clock;
 use solana_program::{
-    account_info::AccountInfo, clock::Epoch, entrypoint::ProgramResult, msg,
-    program_error::ProgramError, program_pack::Pack, pubkey::Pubkey, rent::Rent, sysvar::Sysvar,
+    account_info::AccountInfo,
+    borsh::try_from_slice_unchecked,
+    clock::Clock,
+    clock::Epoch,
+    entrypoint::ProgramResult,
+    msg,
+    program_error::ProgramError,
+    program_memory::sol_memcmp,
+    program_pack::Pack,
+    program_pack::Sealed,
+    pubkey::{Pubkey, PUBKEY_BYTES},
+    rent::Rent,
+    sysvar::Sysvar,
 };
 use spl_token::state::Mint;
 
+use crate::big_vec::BigVec;
 use crate::error::LidoError;
 use crate::logic::get_reserve_available_balance;
 use crate::metrics::Metrics;
 use crate::processor::StakeType;
 use crate::token::{self, Lamports, Rational, StLamports};
 use crate::util::serialize_b58;
 use crate::{
-    account_map::{AccountMap, AccountSet, EntryConstantSize, PubkeyAndEntry},
     MINIMUM_STAKE_ACCOUNT_BALANCE, MINT_AUTHORITY, RESERVE_ACCOUNT, STAKE_AUTHORITY,
+    VALIDATOR_STAKE_ACCOUNT, VALIDATOR_UNSTAKE_ACCOUNT,
 };
-use crate::{VALIDATOR_STAKE_ACCOUNT, VALIDATOR_UNSTAKE_ACCOUNT};
 
 pub const LIDO_VERSION: u8 = 1;
 
 /// Size of a serialized `Lido` struct excluding validators and maintainers.
 ///
 /// To update this, run the tests and replace the value here with the test output.
 pub const LIDO_CONSTANT_SIZE: usize = 353;
-pub const VALIDATOR_CONSTANT_SIZE: usize = 49;
 
-pub type Validators = AccountMap<Validator>;
+/// Types of list entries
+/// Uninitialized should always be a first enum field as it catches empty list data errors
+#[derive(Clone, Debug, PartialEq, Eq, BorshDeserialize, BorshSerialize, Serialize, BorshSchema)]
+pub enum ListEntryType {
+    /// If the account has not been initialized, the enum will be 0
+    Uninitialized,
+    Validator,
+    Maintainer,
+}
+
+impl Default for ListEntryType {
+    fn default() -> Self {
+        ListEntryType::Uninitialized
+    }
+}
+
+/// Storage list for accounts in the pool.
+/// It is used to serialize account list on stake pool initialization
+#[repr(C)]
+#[derive(
+    Clone, Debug, Default, PartialEq, BorshDeserialize, BorshSerialize, BorshSchema, Serialize,
+)]
+pub struct AccountList<T> {
+    /// Data outside of the list, separated out for cheaper deserializations
+    #[serde(skip_serializing)]
+    pub header: ListHeader<T>,
+
+    /// List of accounts in the pool
+    pub entries: Vec<T>,
```

Comment:
`pub entries: Vec<T>` does not hold a vector of pointers, `T` is a `Maintainer` or a `Validator`, which are regular structs. `AccountList<T>` is used in `Initialize` instruction once and then in CLI and tests.

Reply:
Hmm, can we remove it then? During initialization the vec should be empty, so we can just write the zero len manually.

Reply:
I would not advice to remove. If we do than BigVec will leak out to CLI and tests. It will be so inconvenient to stick to data on stack for BigVec: `struct BigVec<'data, T>`. It is much more convenient to just copy data to a regular Vec. I know that I can iter BigVec and collect it to Vec. But it means I should also reimplement all methods `AccountList` has for BigVec. Because when you will copy BigVec to Vec it should have methods `AccountList` had. So, no way